Buffer.php

```diff
@@ -0,0 +1,244 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+namespace Hoa\Iterator;
+
+/**
+ * Class \Hoa\Iterator\Buffer.
+ *
+ * Buffer iterator: Can go backward up to a certain limit, and forward.
+ *
+ * @copyright  Copyright © 2007-2016 Hoa community
+ * @license    New BSD License
+ */
+class          Buffer
+    extends    IteratorIterator
+    implements Outer
+{
+    /**
+     * Buffer key index.
+     *
+     * @const int
+     */
+    const BUFFER_KEY   = 0;
+
+    /**
+     * Buffer value index.
+     *
+     * @const int
+     */
+    const BUFFER_VALUE = 1;
+
+    /**
+     * Current iterator.
+     *
+     * @var \Iterator
+     */
+    protected $_iterator   = null;
+
+    /**
+     * Buffer.
+     *
+     * @var \SplDoublyLinkedList
+     */
+    protected $_buffer     = null;
+
+    /**
+     * Maximum buffer size.
+     *
+     * @var int
+     */
+    protected $_bufferSize = 1;
+
+
+
+    /**
+     * Construct.
+     *
+     * @param   \Iterator  $iterator      Iterator.
+     * @param   int        $bufferSize    Buffer size.
+     * @return  void
+     */
+    public function __construct(\Iterator $iterator, $bufferSize)
+    {
+        $this->_iterator   = $iterator;
+        $this->_bufferSize = max($bufferSize, 1);
+        $this->_buffer     = new \SplDoublyLinkedList();
+
+        return;
+    }
+
+    /**
+     * Get inner iterator.
+     *
+     * @return  \Iterator
+     */
+    public function getInnerIterator()
+    {
+        return $this->_iterator;
+    }
+
+    /**
+     * Get buffer.
+     *
+     * @return  \SplDoublyLinkedList
+     */
+    protected function getBuffer()
+    {
+        return $this->_buffer;
+    }
+
+    /**
+     * Get buffer size.
+     *
+     * @return  int
+     */
+    public function getBufferSize()
+    {
+        return $this->_bufferSize;
+    }
+
+    /**
+     * Return the current element.
+     *
+     * @return  mixed
+     */
+    public function current()
+    {
+        $current = $this->getBuffer()->current();
+
+        return $current[self::BUFFER_VALUE];
+    }
+
+    /**
+     * Return the key of the current element.
+     *
+     * @return  mixed
+     */
+    public function key()
+    {
+        $current = $this->getBuffer()->current();
+
+        return $current[self::BUFFER_KEY];
+    }
+
+    /**
+     * Move forward to next element.
+     *
+     * @return  void
+     */
+    public function next()
+    {
+        $innerIterator = $this->getInnerIterator();
+        $buffer        = $this->getBuffer();
+
+        $buffer->next();
+
+        // End of the buffer, need a new value.
+        if (false === $buffer->valid()) {
+            for (
+                $bufferSize        = count($buffer),
```

Comment:
Hmm, actually using `count` here doesn't affect performance, it is only affected when used as the criteria argument, so the next line might be an issue.